staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/defaulting/algorithm.go

```diff
@@ -34,20 +39,26 @@ func Default(x interface{}, s *structuralschema.Structural) {
 			if prop.Default.Object == nil {
 				continue
 			}
-			if _, found := x[k]; !found {
+			if _, found := x[k]; !found || isNonNullableNull(x[k], &prop) {
```

Comment:
this looks wrong. We have to drop non-nullable nulls in a step before defaulting. This normalization is independent from defaulting (following the KEP). Then this case here should not show up at all.

Reply:
Actually, the reason I do that is that we won't drop non-nullable nulls in lists and we'll default these.

Reply:
Also, yeah, I haven't implemented the "drop non-nullable nulls" yet, that's coming, this is still wip :-)

Reply:
:+1:

Would add that here in the PR, not separately. Both steps are too related.

Reply:
Yeah, I wass planning on doing it in the same PR, that's why it's a wip. it's missing higher-level tests too.